Shadows

lilleyse opened the pull request:
For #2594 

This PR adds support for different types of shadows including cascaded directional shadows, spot light shadows, and point light shadows. It also adds the concept of derived commands which are used for OIT, shadows, and later 3D Tiles.

By default the scene contains a shadow map whose light source is the sun. It is disabled by default, but can be enabled with `scene.shadowMap.enabled = true`. 

Shadow default settings:

| Type  | Casts Shadows | Receives Shadows |
| --- | --- | --- |
| Globe  | False | True |
| Primitives  | False | False |
| Models  | True | True |

Any of these can change by setting `castShadows` or `receiveShadows`.

TODO
* [x] Unit tests
* [ ] Cast shadows for terrain tiles that are out of view: on the [shadows-globe](https://github.com/AnalyticalGraphicsInc/cesium/tree/shadows-globe) branch.
* [x] Figure out the best approach approach for shadows when the light source is below the horizon. Currently if the globe casts shadows, the surface will be completely shadowed and dark. Maybe day-night shading should be on if shadows are on.
* [ ] Possible support for `GroundPrimitive` if needed.
* [x] Integration with higher level constructs like `Viewer`.
* [ ] 2D and Columbus view support.
* [ ] Screenshots for twitter
* [ ] Tech blog post
* [ ] Tutorial, including use cases
* [x] Fix IE11 and Edge - no glPolygonOffset support

Testing TODO
* [x] Review [ShadowMap.js](https://github.com/AnalyticalGraphicsInc/cesium/blob/shadows/Source/Scene/ShadowMap.js)
* [x] Step through in debugger
* [x] FPS with shadows on/off to get a ballpark
* [x] Retest aircraft and satellite CZMLs
* [x] Test with lots of 3D models
* [x] Terrain

TODO now or later
* [ ] Use a good LOD when selecting tiles only in the shadow map frustum, e.g., not SSE from the light's perspective
* [ ] Do not request imagery for tiles only in the shadow map frustum

pjcozzi commented:
Can we make a small Sandcastle example for shadows?  Perhaps just a few lines of code that show choice terrain locations and models.

pjcozzi commented:
> Primitives do neither by default, but they can be enabled.

I thought we already did this?  If not, we definitely want it on the roadmap.  Anything particularly challenging?

lilleyse commented:
> I thought we already did this? If not, we definitely want it on the roadmap. Anything particularly challenging?

I'll change the wording, they do work, but cast and receive are off by default.

pjcozzi commented:
> scene.shadowMap.enabled = true

Perhaps `scene.showShadows = true` or `scene.enableShadows = true` is better.  Do end users need to know anything about the shadow map or that shadows are even based on shadow maps?

pjcozzi commented:
>  Possible support for GroundPrimitive if needed.

Probably not needed since they are modulated with the terrain, which will already be shadowified.

pjcozzi commented:
Or maybe this, https://github.com/AnalyticalGraphicsInc/cesium/pull/3856#issuecomment-210611378, happens here:

> Integration with higher level constructs like Viewer

pjcozzi commented:
>> I thought we already did this? If not, we definitely want it on the roadmap. Anything particularly challenging?
> I'll change the wording, they do work, but cast and receive are off by default.

Ah.  Remind me, why are they off by default?  Shouldn't we treat them like models?

lilleyse commented:
> Ah. Remind me, why are they off by default? Shouldn't we treat them like models?

In a lot of cases Primitive aren't "real world" entities, so users may not want them shadowed.

pjcozzi commented:
Is the roadmap in #2594 up-to-date?

pjcozzi commented:
> In a lot of cases Primitive aren't "real world" entities, so users may not want them shadowed.

OK will see how it goes.  I would almost want to put the default on the geometry, not `Primitive`, but that is misplaced and would not work well with heterogeneous batching.

pjcozzi commented on Apps/Sandcastle/gallery/development/Multiple Shadows.html:
```diff
@@ -0,0 +1,132 @@
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no">
+    <meta name="description" content="Shadow maps.">
+    <meta name="cesium-sandcastle-labels" content="Tutorials,Showcases">
```
Both development Sandcastle examples have the wrong labels.

lilleyse commented:
> Is the roadmap in #2594 up-to-date?

I need to add a few new things to it, but it's basically up-to-date.

pjcozzi commented:
This would be for the roadmap, but what do you think of dynamically resizing shadow maps based on the window size (but would need to wait for it to stabilize, e.g., for when someone drags to resize).  For example, here's what the default view in the Sandcastle example looks like in full screen:

![image](https://cloud.githubusercontent.com/assets/782098/14573624/21db0c9c-0324-11e6-8700-2b151969e735.png)

pjcozzi commented:
Did you test this with a satellite model in orbit?  I emailed you one.

lilleyse commented:
I tested with an object far out in space, I'll give your example a try too.

pjcozzi commented:
>  Fix Globe.update which causes flickery terrain tiles currently

Is this also the cause of, for example:

![image](https://cloud.githubusercontent.com/assets/782098/14573831/79d67e80-0325-11e6-804a-5f4e95e4e96a.png)

lilleyse commented:
Most likely all the terrain issues are related.

bagnell commented:
>>Possible support for GroundPrimitive if needed.
>Probably not needed since they are modulated with the terrain, which will already be shadowified.

The translucent ones probably won't, but the opaque ones will. Not sure if opaque `GroundPrimitive`s will ever be used in practice.

pjcozzi commented:
> The translucent ones probably won't, but the opaque ones will. Not sure if opaque GroundPrimitives will ever be used in practice.

Good point.  It's worth adding to the roadmap, but I don't know if it will be asked for.

pjcozzi commented:
Could be for the roadmap if it affects a lot of users, but software shadows can introduce some z-fighting artifacts, e.g.,

Without:

![image](https://cloud.githubusercontent.com/assets/782098/14574085/5744b9f2-0327-11e6-90ed-e04ce056eadf.png)

With:

![image](https://cloud.githubusercontent.com/assets/782098/14574075/4b9f97a2-0327-11e6-9b71-f327b209f1ac.png)

pjcozzi commented:
The shadows disable when I select "Exponential Shadows."  Should they?  Should we remove this option (in Cesium, not just Sandcastle)?

pjcozzi commented:
For the new simple Sandcastle example, I think it should also include:
* Toggle for soft shadows
* Ability to change the shadow map resolution since I expect this will be a fairly common task.  We don't want to go above 1K for the default given the CSM and WebGL limits, right?

pjcozzi commented:
> Could be for the roadmap if it affects a lot of users, but software shadows can introduce some z-fighting artifacts...

I am noticing this a lot.  What do you think is involved?  Just different fudge factors when using soft shadows?

lilleyse commented:
> The shadows disable when I select "Exponential Shadows."

It could be something broke since I last checked them. Right now they should always be off, but if we experiment with different smooth shading techniques exponential shadow maps will be important to have. Maybe a separate branch makes sense here.

lilleyse commented:
> I am noticing this a lot. What do you think is involved? Just different fudge factors when using soft shadows?

I've noticed it too I just haven't put the effort into fixing them since they don't even look that great. For the roadmap though.

lilleyse commented:
> We don't want to go above 1K for the default given the CSM and WebGL limits, right?

2K is well supported.

pjcozzi commented:
> 2K is well supported.

But we can't default to 2K because of the cascades, right?

pjcozzi commented:
> Maybe a separate branch makes sense here.

OK

pjcozzi commented:
Did you use the hosted branch to test on mobile at all?

lilleyse commented:
> But we can't default to 2K because of the cascades, right?

Oh true. 4k is at 75% adoption.

mramato commented:
> Oh true. 4k is at 75% adoption.

Can't we detect this at run time and choose the best default the hardware can provide?

lilleyse commented:
> Can't we detect this at run time and choose the best default the hardware can provide?

Good point. We can definitely do that.

lilleyse commented:
> Did you use the hosted branch to test on mobile at all?

I'll try later when I have better service, but @mramato checked a while ago and it worked.

pjcozzi commented on Apps/Sandcastle/gallery/development/Multiple Shadows.html:
```diff
@@ -0,0 +1,132 @@
+
+        scene.shadowMap = shadowMap1;
+
+        var CustomPrimitive = function(shadowMap) {
```
Might want to comment here that this is a workaround to test multiple shadow maps without Cesium yet supporting multiple lights.

pjcozzi commented on Source/Core/Matrix4.js:
```diff
@@ -958,6 +959,54 @@ define([
+     * @param {Matrix4} result The object in which the result will be stored.
+     * @returns {Matrix4} The modified result parameter.
+     */
+    Matrix4.computeView = function(direction, up, right, position, result) {
```
Is a more common ordering `position, direction, up, right`?

pjcozzi commented on Source/Renderer/CubeMap.js:
```diff
@@ -169,7 +169,7 @@ define([
         this._negativeZ = new CubeMapFace(gl, texture, textureTarget, gl.TEXTURE_CUBE_MAP_NEGATIVE_Z, pixelFormat, pixelDatatype, size, preMultiplyAlpha, flipY);
 
-        this.sampler = new Sampler();
+        this.sampler = defined(options.sampler) ? options.sampler : new Sampler();
```
Good eye, I am surprised we never hit this before.

pjcozzi commented on Source/Renderer/DrawCommand.js:
```diff
@@ -37,6 +37,8 @@ define([
         this._debugShowBoundingVolume = defaultValue(options.debugShowBoundingVolume, false);
         this._debugOverlappingFrustums = 0;
+        this._castShadows = defaultValue(options.castShadows, false);
+        this._receiveShadows = defaultValue(options.receiveShadows, false); // TODO : not used right now, but keep if it proves useful
```
What is the plan with this TODO?

lilleyse replied:
It's being used now, I'll remove the TODO.

pjcozzi commented on Source/Renderer/UniformState.js:
```diff
@@ -857,6 +857,23 @@ define([
+     *
+     * @param {Object} camera The camera to synchronize with.
+     */
+    UniformState.prototype.updateCamera = function(camera) {
```
Does this need to set any other matrices with `view` in them, e.g., MVP, or set any dirty flags?

pjcozzi replied:
Ah, that is all in `setView`.

pjcozzi commented on Source/Renderer/UniformState.js:
```diff
@@ -792,6 +790,7 @@ define([
         uniformState._inverseModelViewDirty = true;
         uniformState._inverseModelView3DDirty = true;
         uniformState._viewProjectionDirty = true;
+        uniformState._inverseViewProjectionDirty = true;
```
Good catch!

pjcozzi commented on Source/Scene/Camera.js:
```diff
@@ -2841,13 +2819,20 @@ define([
+        result._transformChanged = true;
+
+        // Clone frustum only if it is the same type
+        if (defined(camera.frustum.left) === defined(result.frustum.left)) {
```
What's this for?  If a user passed in their own `result` with another frustum type?

lilleyse replied:
Yes that was the reason. The shadows code doesn't depend on this check though. Should I remove this, or expand on the comment?

pjcozzi replied:
I would leave this how it was - or as close to it as possible.

pjcozzi commented on Source/Scene/Globe.js:
```diff
@@ -188,6 +188,22 @@ define([
         this.depthTestAgainstTerrain = false;
 
+        /**
+         * Determines whether the globe will cast shadows when shadow mapping is enabled.
```
Can we beef up this description (and to a lesser extent, the description for `receiveShadows`).
* Be explicit that shadows are cast onto anything with `receiveShadows` set to `true`, not just terrain, and vice versa - shadows can be received by other things.
* This can impact performance since the terrain is rendered a second time from the perspective of the sun/light, which will also load terrain tiles potentially out-side of view.

pjcozzi commented on Source/Scene/GlobeSurfaceTileProvider.js:
```diff
@@ -648,107 +651,123 @@ define([
-
-            minMaxHeight : new Cartesian2(),
-            scaleAndBias : new Matrix4()
+            // make a separate object so that changes to the properties are seen on
```
This is because we want a shallow copy here, right?  Let's make the comment explicit.

In general, we might want to think about this for derived commands since this code needs to know an awful lot about how they work in order to get this right, maybe this is part of a bigger uniform refactor for uniform buffers, auto packing, etc.

pjcozzi commented on CHANGES.md:
```diff
@@ -1,6 +1,12 @@
+### Shadows TODO
+* Added `Primitive.castShadows` and `Primitive.receiveShadows`.
+* Added `Globe.castShadows` and `Globe.receiveShadows`.
+* Added `Model.castShadows` and `Model.receiveShadows`.
```
And the constructor/from functions.

pjcozzi commented on CHANGES.md:
```diff
@@ -1,6 +1,12 @@
 Change Log
 ==========
 
+### Shadows TODO
```
What about properties on `Scene` (haven't looked at these yet, but I assume enable/disable, soft shadows, near/far, etc.).

pjcozzi replied:
I didn't see these, but perhaps that is part of the ` Integration with higher level constructs like Viewer.` task item.

lilleyse replied:
Yeah I think so. All shadow properties like this are inside each `ShadowMap` right now.

pjcozzi replied:
Remove `TODO`, move this to the `1.22` section, and also add a link to the new Sandcastle example (using the future cesiumjs.org base path).

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -539,6 +541,18 @@ define([
          */
         this.fog = new Fog();
 
+        this._sunCamera = new Camera(this);
```
Is this member really needed?  It is only used in one place after the constructor.

lilleyse replied:
It is needed for now, but only because someone could change `Scene.shadowMap` to something else, like in the Sandcastle.

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -539,6 +541,18 @@ define([
+            context : context,
+            lightCamera : this._sunCamera
+        });
+        this.shadowMap.enabled = false;
```
Can `enabled` be part of `options` passed to the constructor function?

pjcozzi commented on Source/Scene/ShadowMapShader.js:
```diff
@@ -0,0 +1,368 @@
+    function ShadowMapShader() {
+    }
+
+    function findVarying(shaderSource, names) {
```
Should this and helpers below be static functions on `ShaderSource` like https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Renderer/ShaderSource.js#L287-L290?

pjcozzi replied:
For helpers, I just mean `findNormalVarying` and `findPositionVarying`, not the shadow-map-specific functions, which are well placed here.

lilleyse replied:
Either way is fine with me, I made the change.

pjcozzi commented on Source/Scene/ShadowMapShader.js:
```diff
@@ -0,0 +1,368 @@
+                'void main() \n' +
+                '{ \n' +
+                '    czm_shadow_cast_main(); \n' +
+                '    v_positionEC = (czm_inverseProjection * gl_Position).xyz; \n' +
```
Offline, did you say you had better precision luck with this instead of going directly from model/world/RTE/whatever?  Could be a research item for the roadmap since I feel like this should be worse.

lilleyse replied:
I think when we talked about that it was related to the Globe shader which wasn't outputting as precise `v_positionEC`, but I since fixed that.

pjcozzi commented on Source/Scene/ShadowMapShader.js:
```diff
@@ -0,0 +1,368 @@
+            if (!hasPositionVarying) {
+                fsSource += 'varying vec3 v_positionEC; \n';
+            }
+            fsSource += 'uniform vec4 u_shadowMapLightPositionEC; \n';
```
`u_` is fine.  For the added uniforms for the batch table in 3D Tiles, I am using `tile_`, which has worked out well compared to `czm_`.  If you think it is a useful distinction, perhaps `shadow_` would be worthwhile here.

pjcozzi replied:
Maybe things become even more concise `u_shadowMapTexture` -> `shadowMap_texture`.  I dunno.

pjcozzi commented on Source/Scene/ShadowMapShader.js:
```diff
@@ -0,0 +1,368 @@
+            fsSource +=
+                'void main() \n' +
+                '{ \n' +
+                '    czm_shadow_cast_main(); \n' +
```
If you don't already have it on the roadmap, please add using a pass-through fragment shader when we know we can (I want to do the same for picking) since I doubt drivers optimize the shader when we disable color writes and many shaders will not need to discard here.

pjcozzi replied:
Just a reminder in case this isn't already on the roadmap.

pjcozzi commented on Source/Scene/ShadowMapShader.js:
```diff
@@ -0,0 +1,368 @@
+
+        var fsSource =
+            'uniform mat4 u_shadowMapMatrix; \n' +
+            'uniform vec3 u_shadowMapLightDirectionEC; \n' +
```
Can we combine these `vec3` and `vec4` uniforms into one or more matrix uniforms then "unpack" here?  Or did we discuss this in a previous PR?

pjcozzi commented on Source/Scene/ShadowMapShader.js:
```diff
@@ -0,0 +1,368 @@
+            'uniform vec4 u_shadowMapLightPositionEC; \n' +
+            'uniform vec4 u_shadowMapNormalOffsetScaleDistanceMaxDistanceAndDarkness; \n' +
+            'uniform vec4 u_shadowMapTexelSizeDepthBiasAndNormalShadingSmooth; \n' +
+            'uniform sampler2D u_shadowMapTexture; \n' +
```
Should we check `isPointLight` and then include just the right sampler so we're not relying on the compiler?

pjcozzi commented on Source/Scene/ShadowMapShader.js:
```diff
@@ -0,0 +1,368 @@
+            '{ \n' +
+            (bias.normalOffset && hasNormalVarying ?
+            '    float normalOffset = u_shadowMapNormalOffsetScaleDistanceMaxDistanceAndDarkness.x; \n' +
+            '    // Offset the shadow position in the direction of the normal for perpendicular and back faces \n' +
```
Do you think it is better to make this a JavaScript comment so it gets syntax highlighting?

pjcozzi replied:
Same comment below.

pjcozzi commented:
ShadowMap.js -

```
        this._numberOfPasses = numberOfPasses;
        this._passCameras = new Array(numberOfPasses);
        this._passStates = new Array(numberOfPasses);
        this._passFramebuffers = new Array(numberOfPasses);
        this._passTextureOffsets = new Array(numberOfPasses);
        this._passCommands = new Array(numberOfPasses);
        this._passCullingVolumes = new Array(numberOfPasses);
```

Is a AOS better than SOA here?

pjcozzi commented:
ShadowMap.js -

> this._usesCubeMap = true;

This is always `true`, should we remove it?

pjcozzi commented:
ShadowMap.js -

> this._radius = defaultValue(options.radius, 100.0);

Is this better named `_pointLightRadius` since it is more precise in the shadow map context?

pjcozzi commented:
ShadowMap.js -

> enabled : true, // TODO : need to handle objects that don't use back face culling, like walls

What is the plan for this TODO?  Perhaps derived commands now make this possible?

pjcozzi commented:
ShadowMap.js -

> ShadowMap.prototype._createRenderStates

Maybe name this:
```
/**
 * @private
 */
ShadowMap.prototype.debugCreateRenderStates
```

pjcozzi commented:
ShadowMap.js -

> '    if (uv.y < 0.5) { \n' +

In GLSL, the style is to put the `{` on a new line.

pjcozzi commented:
ShadowMap.js -

> debugCommand = shadowMap._debugShadowViewCommand = createDebugShadowViewCommand(shadowMap, context);

Some folks do this, but I would split this into two lines as I've seen people get mixed up by this, e.g., not realize there is two assignments at first glance.

pjcozzi commented:
ShadowMap.js -

> function createDebugFrustum(camera, color)

How hard is it to separate this out into general debug code (`@private`); it will be useful for 3D Tiles.

pjcozzi commented:
ShadowMap.js -

> var maximumDistances = [50.0, 300.0, Number.MAX_VALUE, Number.MAX_VALUE];

How did you come up with 50 and 300?  Do these need to be tunable?  In practice, would anyone want to change them?

pjcozzi commented:
ShadowMap.js -

> // TODO : This is just an estimate, should take scene geometry into account

Delete this TODO or move it to a research idea in the roadmap?

pjcozzi commented:
> Integration with higher level constructs like Viewer.

This also needs to be wired into the Entity API, perhaps for individual entities, not each sub-object.

pjcozzi commented:
ShadowMap.js -

> if (isTerrain) {

Do we have a roadmap item to make this less coupled?

pjcozzi commented:
@lilleyse did you test this with VR to make sure it plays well with Cesium's two viewports?

pjcozzi commented:
Is the `NaN` for the last cascade distance expected?

![image](https://cloud.githubusercontent.com/assets/782098/14604104/d3a191fe-053f-11e6-899b-f7e8e8f441b5.png)

pjcozzi commented:
ShadowMap.js -

> // Limit min to 0.0. Sometimes precision errors cause it to go slightly negative, which affects the frustum extent computations below.

This is in NDC, right?  Should we also clamp max to 1.0?

pjcozzi commented:
ShadowMap.js -

```javascript
cascadeCamera.frustum.left = left + min.x * (right - left);
// ...
```
would be better written as
```javascript
var frustum = cascadeCamera.frustum;
frustum.left = left + min.x * (right - left);
// ...
```

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -1851,12 +2012,40 @@ define([
+        scene._groundPrimitives.update(frameState);
+        scene._primitives.update(frameState);
+
+        var shadowMaps = frameState.shadowMaps;
```
Maybe move this new code into its own function so all lines in this function are at a similar level of abstraction.

https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/CodingGuide#functions

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -1851,12 +2012,40 @@ define([
+            shadowMap.update(frameState);
+
+            if (!shadowMap.outOfView && defined(scene._globe) && scene._globe.castShadows) {
+                var sceneCamera = frameState.camera;
```
Instead of save/set/restore, is there a cleaner way to do this?  Perhaps create a separate frame state for the shadow map?

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -1851,12 +2012,40 @@ define([
+                frameState.commandList.length = 0;
+
+                // Collect terrain commands from the light's POV
+                scene._globe.update(frameState);
```
`update` is called twice for the globe, once as usual and once for the shadow map, but it looks like `update` is only called once for primitives (or not?).  Can we add a comment about why?

pjcozzi replied:
Ah, I see the commands being added in `executeShadowMapCommands` from the two different command lists.  Do primitives outside of the view frustum cast shadows into the view frustum?  Won't they be culled?

lilleyse replied:
Yeah primitives only need to be updated once since they aren't culled per-camera like globe tiles, I'll add a comment.

lilleyse replied:
Unless I'm wrong about that.. it's true for models at least.

pjcozzi replied:
`Model` and `Primitive` are treated the same by `Scene`.

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -1659,6 +1716,108 @@ define([
         }
     }
 
+    function insertShadowCommands(scene, commandList, insertAll, shadowMap, shadowPassCommands) {
```
For the roadmap, do you think it is worthwhile to consider doing all the culling (view frustum and shadow maps) at once instead of distributing it between createPVS and here?

pjcozzi replied:
Likewise, when commands come from a spatial data structure - like terrain and 3D Tiles - could these checks be any more efficient?

lilleyse replied:
> For the roadmap, do you think it is worthwhile to consider doing all the culling (view frustum and shadow maps) at once instead of distributing it between createPVS and here?

I can add that to the roadmap. The main reason we didn't do that is createPVS is really tied to the multifrustum binning, which doesn't apply for shadow commands.

lilleyse replied:
> Likewise, when commands come from a spatial data structure - like terrain and 3D Tiles - could these checks be any more efficient?

Hm possibly, I'll add that too.

pjcozzi commented:
shadowMap.js - 

> castUniformMap = combineUniforms(shadowMap, command.uniformMap, isTerrain);

Passing `isTerrain` deserves a comment here.

pjcozzi commented:
> Passing isTerrain deserves a comment here.

Nevermind I thought it was calling `combine`.

pjcozzi commented:
> return Cartesian4.fromElements(texelStepSize.x, texelStepSize.y, bias.depthBias, bias.normalShadingSmooth, scratchUniformCartesian1);

Using a scratch here is OK now.  We might do this with uniforms in one or two other places, but it is dangerous since it assumes that the return value of the function is used immediately and not queued up by the renderer.  Generally, I _think_ we use members on the uniform map for scratch here.

pjcozzi commented:
shadowMap.js -

> result.count = skirtIndex;

Add a comment why.

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -1659,6 +1716,108 @@ define([
+        }
+    }
+
+    function executeShadowMapCommands(scene) {
```
Perhaps this is better named precisely as `executeShadowMapCastCommands`.

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -1344,6 +1396,8 @@ define([
 
         if (scene.debugShowCommands || scene.debugShowFrustums) {
             executeDebugCommand(command, scene, passState);
+        } else if (scene.frameState.shadowMaps.length > 0 && command.receiveShadows && defined(command.derivedCommands.shadows.receiveCommand)) {
```
Is the last check, `defined(command.derivedCommands.shadows.receiveCommand)`, needed?

pjcozzi commented:
That's all my comments for this pass.

lilleyse commented:
> How hard is it to separate this out into general debug code (@private); it will be useful for 3D Tiles.

This function should be easy to use anywhere. The only caveat is that it must be called every frame if the camera changes (not a problem in freeze frame mode).

lilleyse commented:
> How did you come up with 50 and 300? Do these need to be tunable? In practice, would anyone want to change them?

I started at a view where I thought the aliasing was acceptable, and just printed out the values. Then I just tweaked them after trying different views and models. Too small and the third and fourth cascade would suffer, too large and the shadows would look too blocky up close. At the very least I could turn `maximumDistances` into a private member variable.

lilleyse commented:
> Delete this TODO or move it to a research idea in the roadmap?

It's also in the road map, I removed the TODO

lilleyse commented:
> Do we have a roadmap item to make this less coupled?

Yes

lilleyse commented:
> @lilleyse did you test this with VR to make sure it plays well with Cesium's two viewports?

Not yet.

lilleyse commented:
> This is in NDC, right? Should we also clamp max to 1.0?

Heh, @bagnell asked a similar question in a previous PR. Basically being negative is the real problem here and the values slightly above 1 aren't a problem. I'll clamp to [0, 1] since it is clearer that way.

lilleyse commented:
The VR mode is close, but there are some aliasing issues.

pjcozzi commented:
>> How hard is it to separate this out into general debug code (@private); it will be useful for 3D Tiles.

> This function should be easy to use anywhere. The only caveat is that it must be called every frame if the camera changes (not a problem in freeze frame mode).

Yes, would not be a problem for 3D Tiles.

mramato commented:
1. Starting with shadows disabled seems to make it impossible to enable them (probably because of cached draw commands?).

2. It's really easy to crash Cesium by starting with shadows enabled and then toggling between disabled/enabled as new data is coming it.

mramato commented:
Shadows seem to be broken in IE and Edge (may be 3d-tiles related only).  In IE, everything just goes black.  In Edge, I get this:

![image](https://cloud.githubusercontent.com/assets/1495196/14937861/ee98b3f6-0ee1-11e6-8be7-465b0c0df694.png)

lilleyse commented:
> Shadows seem to be broken in IE and Edge (may be 3d-tiles related only). In IE, everything just goes black. In Edge, I get this:

We talked about this offline a little, but it looks like polygonOffset isn't supported in IE11, and probably Edge too (I'll check later). I'm using that for preventing shadow aliasing at sharp angles, and there may be other ways to get around it. Besides that, it works for me in IE including the 3d-tiles-shadows branch. I'm not sure why it's completely black for you.

mramato commented:
> I'm not sure why it's completely black for you.

In IE on my work desktop (Nvidia/Windows 7), I see the same thing as Edge (the above picture).  In IE on my laptop (Intel HD4000/Windows 10) I see the black (though I haven't tried in a while.

I also noticed that there is some sort of weird race condition/initialization problems, even in Chrome.  Go to https://cesiumjs.org/NewYork/index.html?shadows and don't touch the mouse, the image will be super-dark until the first mouse interaction.

lilleyse commented:
> I also noticed that there is some sort of weird race condition/initialization problems, even in Chrome. Go to https://cesiumjs.org/NewYork/index.html?shadows and don't touch the mouse, the image will be super-dark until the first mouse interaction.

Should be fixed now.

lilleyse commented:
> 1. Starting with shadows disabled seems to make it impossible to enable them (probably because of cached draw commands?).
2. It's really easy to crash Cesium by starting with shadows enabled and then toggling between disabled/enabled as new data is coming it.

This is fixed now. The derived commands only updated if they were dirty, but didn't take into account if the shadow maps had changed (e.g. turned on/off, soft shadows, and any other settings). I borrowed the idea of keeping a `lastStyleTime` from 3d-tiles for checking when a command needs to create new derived commands. This also improves the Sandcastle demo a lot because it no longer needs to destroy the globe whenever you tweak shadow parameters.

lilleyse commented:
@mramato What do you think is the best approach for adding `castShadows` and `receiveShadows` to the Entity API? Right now it looks like this for models:

```
entity = viewer.entities.add({
            position : position,
            orientation : orientation,
            model : {
                uri : currentModel.url,
                castShadows : true,
                receiveShadows : true
            }
        });
```

I can do the same thing for every other entity type. Also I will need to update the `GeometryVisualizer` to handle batching, and update the CZML code as well. But before I get too far into this I want to know if this is the right path, and if there's anything I'm missing.

mramato commented:
> What do you think is the best approach for adding castShadows and receiveShadows to the Entity API?

That's a good question. What you are doing is definitely in line with how we usually handle this.  The only other possibility would be to have a single `castShadows`/`receiveShadows` for each entity rather than on the individual graphics objects; but I think your options is better for consistency.

mramato commented:
@lilleyse should IE 11 be working now?  I still have heavy visual artifacts for Simple City:

![image](https://cloud.githubusercontent.com/assets/1495196/15575139/5b331726-231f-11e6-9c08-c606deab9860.png)

lilleyse commented:
> @lilleyse should IE 11 be working now? I still have heavy visual artifacts for Simple City:

I haven't fixed that issue yet, hopefully I'll find a good trade-off soon.

mramato commented:
> I haven't fixed that issue yet, hopefully I'll find a good trade-off soon.

No worries, I just didn't see a checkbox for it (Might want to add one) so I wasn't sure.  Thanks

pjcozzi commented:
Is this ready for final testing?

lilleyse commented:
There is one breaking bug still that happens when zooming too far out, related to derived commands that I recently introduced. But if you want to test models and terrain it should be good to go.

IE has a lot of failing tests currently despite seeming fine in the demo. It seems that just rendering into the shadow map causes the test to fail, regardless of whether the test is even checking that.

mramato commented:
FYI, I can reliably crash the Sandcastle demo in Edge but switching to the wooden tower and animating at 14400.

lilleyse commented:
> FYI, I can reliably crash the Sandcastle demo in Edge but switching to the wooden tower and animating at 14400.

Seems to happening in Chrome as well, and is being caused by the same bug. I'm tempted to just change the code back to the way it was before with a single-line change.

lilleyse commented:
I moved the globe update code to [shadows-globe] (https://github.com/AnalyticalGraphicsInc/cesium/tree/shadows-globe) and fixed the aliasing issues in IE as well as the recent versions of Chrome and Firefox. When #3973 is in, this should be good to go.

pjcozzi commented:
@lilleyse if you haven't already, can you please document that currently only terrain/models/primitives/etc in view cast shadows.

pjcozzi commented:
@lilleyse can you update the task list in this PR and the shadows roadmap?  I will take care of the **Testing TODO** section now.

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -548,6 +551,18 @@ define([
+         * Render shadows in the scene.
+         * @type {ShadowMap}
+         */
+        this.shadowMap = new ShadowMap({
```
This property is not in CHANGES.md and the reference doc is a bit lacking.

pjcozzi commented:
When running the Sandcastle example in Mac/Chrome, I get
```
[.CommandBufferContext]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glClear: framebuffer incomplete (clear)
```
The screen also flashes black sometimes, for example, during a camera flight.

pjcozzi commented on Specs/Scene/ShadowMapSpec.js:
```diff
@@ -0,0 +1,1072 @@
+    var boxHeight = 4.0;
+    var floorHeight = -1.0;
+
+    var boxUrl = './Data/Models/Shadows/Box.gltf';
```
Are all the test models in the right location now?  At one time, we had Sandcastle examples pointing to the tests or vice versa.

lilleyse replied:
That's still the case for the test 3d-tiles data, but it's fine here.

pjcozzi replied:
ah, right!

pjcozzi commented:
Using the development Sandcastle example and the Shadow Tester 2 model, I noticed that disabling/enabling the normal offset shifts the shadow's location (on terrain, but not the model/primitive self-shadow).

![ok](https://cloud.githubusercontent.com/assets/782098/15628023/b55020be-24c3-11e6-8d71-0858babedaec.gif)

Is the default value too aggressive?  It's important that the shadows be accurate, but would tuning this value down create more shadow acne?

pjcozzi commented on Source/DataSources/CzmlDataSource.js:
```diff
@@ -1185,6 +1185,8 @@ define([
         processPacketData(Number, model, 'scale', modelData.scale, interval, sourceUri, entityCollection);
         processPacketData(Number, model, 'minimumPixelSize', modelData.minimumPixelSize, interval, sourceUri, entityCollection);
         processPacketData(Boolean, model, 'incrementallyLoadTextures', modelData.incrementallyLoadTextures, interval, sourceUri, entityCollection);
+        processPacketData(Boolean, model, 'castShadows', modelData.castShadows, interval, sourceUri, entityCollection);
```
Did you update the CZML reference doc and czml-writer project?  If not, put it on the roadmap, and let's do it soon.

lilleyse replied:
I haven't it's on the roadmap now.

pjcozzi commented on Source/Renderer/DrawCommand.js:
```diff
@@ -37,7 +37,10 @@ define([
+        this._castShadows = defaultValue(options.castShadows, false);
+        this._receiveShadows = defaultValue(options.receiveShadows, false);
         this._dirty = true;
+        this._lastDirtyTime = 0;
```
This shouldn't start with `_` since it is not private.

lilleyse replied:
`_dirty` is similar. I can change both.

pjcozzi commented on Source/Scene/QuadtreePrimitive.js:
```diff
@@ -505,7 +505,8 @@ define([
 
     function screenSpaceError(primitive, frameState, tile) {
-        if (frameState.mode === SceneMode.SCENE2D) {
+        // TODO: check for sseDenominator so shadow cameras with orthographic frustum works. check correctness.
```
What's the plan with this TODO?  Roadmap?

lilleyse replied:
Definitely relevant for the shadows-globe branch, but I can remove it from this.

pjcozzi commented on Source/Scene/ShadowMap.js:
```diff
@@ -0,0 +1,1648 @@
+         * @default false
+         * @private
+         */
+        dirty : {
```
This could just be a regular property without get/set functions.

pjcozzi commented:
I submitted a few trivial style tweaks.

This is so close!

lilleyse commented:
> [.CommandBufferContext]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glClear: framebuffer incomplete (clear)

Do you also see this error when running the tests?

lilleyse commented:
> Using the development Sandcastle example and the Shadow Tester 2 model, I noticed that disabling/enabling the normal offset shifts the shadow's location (on terrain, but not the model/primitive self-shadow).

Unfortunately this is a side effect of that technique. In the demo you can switch between bias modes, so if you turned it off/on for primitives you would notice the same thing. At some point these values worked well, but I think I can safely reduce them a bit now.

pjcozzi commented:
> Do you also see this error when running the tests?

Yes, when running http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/Specs/SpecRunner.html?spec=Scene%2FShadowMap

I get 27 of
> SpecRunner.html:1 [.CommandBufferContext.Offscreen-MainThread-0x7ff8bd881150]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glClear: framebuffer incomplete (clear)

pjcozzi commented:
> At some point these values worked well, but I think I can safely reduce them a bit now.

Sounds good.

Also, if the shadow shift is still noticeable, please document it so users that need ultra precision shadows can tune it.

lilleyse commented:
Updated. I don't have my laptop with me so it's hard to investigate the OSX/Chrome error...

pjcozzi commented on Source/Scene/ShadowMap.js:
```diff
@@ -179,19 +179,22 @@ define([
         }
         this._polygonOffsetSupported = polygonOffsetSupported;
 
+        // The normalOffset bias pushes the shadows forward slightly, and may be disabled
```
Put this in the reference doc so users will see it.

pjcozzi commented:
@lilleyse what do you suggest?  Wait to merge until we figure out the Mac issue?  Or merge now to increase the chances of more people trying it out in master?

lilleyse commented:
Updated. I was able to test it out and fix the problem.

pjcozzi commented:
The WebGL error is gone, but I have one new shadow map test failure:
```
Ran 34 of 7358 specs - run all
34 specs, 1 failure
Spec List | Failures
Scene/ShadowMap directional shadow map
Expected [ 72, 72, 72, 255 ] to equal [ 240, 240, 240, 255 ].
Error: Expected [ 72, 72, 72, 255 ] to equal [ 240, 240, 240, 255 ].
    at stack (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/ThirdParty/jasmine-2.2.0/jasmine.js:1475:17)
    at buildExpectationResult (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/ThirdParty/jasmine-2.2.0/jasmine.js:1445:14)
    at Spec.expectationResultFactory (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/ThirdParty/jasmine-2.2.0/jasmine.js:584:18)
    at Spec.addExpectationResult (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/ThirdParty/jasmine-2.2.0/jasmine.js:332:34)
    at Expectation.addExpectationResult (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/ThirdParty/jasmine-2.2.0/jasmine.js:528:21)
    at Expectation.toEqual (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/ThirdParty/jasmine-2.2.0/jasmine.js:1399:12)
    at verifyShadows (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/Specs/Scene/ShadowMapSpec.js:373:26)
    at http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/Specs/Scene/ShadowMapSpec.js:623:9
    at Object.<anonymous> (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/Specs/spec-main.js:139:30)
    at attemptAsync (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/ThirdParty/jasmine-2.2.0/jasmine.js:1793:24)
Expected [ 72, 72, 72, 255 ] to equal [ 240, 240, 240, 255 ].
Error: Expected [ 72, 72, 72, 255 ] to equal [ 240, 240, 240, 255 ].
    at stack (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/ThirdParty/jasmine-2.2.0/jasmine.js:1475:17)
    at buildExpectationResult (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/ThirdParty/jasmine-2.2.0/jasmine.js:1445:14)
    at Spec.expectationResultFactory (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/ThirdParty/jasmine-2.2.0/jasmine.js:584:18)
    at Spec.addExpectationResult (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/ThirdParty/jasmine-2.2.0/jasmine.js:332:34)
    at Expectation.addExpectationResult (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/ThirdParty/jasmine-2.2.0/jasmine.js:528:21)
    at Expectation.toEqual (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/ThirdParty/jasmine-2.2.0/jasmine.js:1399:12)
    at verifyShadows (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/Specs/Scene/ShadowMapSpec.js:385:26)
    at http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/Specs/Scene/ShadowMapSpec.js:623:9
    at Object.<anonymous> (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/Specs/spec-main.js:139:30)
    at attemptAsync (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/shadows/ThirdParty/jasmine-2.2.0/jasmine.js:1793:24)
```
The blinking also appears to be fixed, but I thought I saw it one time, but I can't reproduce it.

pjcozzi commented:
Is #2594 completely up-to-date?  There are still items above (https://github.com/AnalyticalGraphicsInc/cesium/pull/3856#issue-148738323) that are not included.  Is this because you plan to do them before we merge this?

lilleyse commented:
The problem there is most likely a z-fighting issue. I made a small change but I can't verify if it fixes the test. A lot of new aliasing problems like this are caused by changes in the browsers lately. In Chrome 49 polygonOffset seemed to work fine, but now there is no effect when toggling it in the demo.

lilleyse commented:
> Is #2594 completely up-to-date? There are still items above (#3856 (comment)) that are not included. Is this because you plan to do them before we merge this?

I moved the rest of the unfinished items here to the roadmap.

pjcozzi commented:
All looks good now.

I saw the blink again, this time from a stationary view, perhaps due to new terrain tiles loading.  If this is a common issue, we'll track it down.

pjcozzi commented:
Also update CHANGES.md in master to include a link to the Sandcastle demo (using the future cesiumjs.org base path).

lilleyse commented:
> Also update CHANGES.md in master to include a link to the Sandcastle demo (using the future cesiumjs.org base path).

Done.
